algorithms/kernel/dtrees/forest/classification/df_classification_model_builder.cpp

```diff
@@ -51,30 +51,76 @@ services::Status ModelBuilder::initialize(size_t nClasses, size_t nTrees)
     modelImplRef.resize(nTrees);
     modelImplRef._impurityTables.reset();
     modelImplRef._nNodeSampleTables.reset();
-    modelImplRef._probTbl.reset();
     modelImplRef._nTree.set(nTrees);
     return Status();
 }
 
-services::Status ModelBuilder::createTreeInternal(size_t nNodes, TreeId& resId)
+services::Status ModelBuilder::createTreeInternal(const size_t nNodes, TreeId & resId)
 {
-    decision_forest::classification::internal::ModelImpl& modelImplRef = daal::algorithms::dtrees::internal::getModelRef<decision_forest::classification::internal::ModelImpl,ModelPtr>(_model);
-    return daal::algorithms::dtrees::internal::createTreeInternal(modelImplRef._serializationData, nNodes, resId);
+    decision_forest::classification::internal::ModelImpl & modelImplRef =
+        daal::algorithms::dtrees::internal::getModelRef<decision_forest::classification::internal::ModelImpl, ModelPtr>(_model);
+    services::Status status = daal::algorithms::dtrees::internal::createTreeInternal(modelImplRef._serializationData, nNodes, resId);
+    if (status.ok())
+    {
+        const auto probTbl = new HomogenNumericTable<double>(nNodes, _nClasses, NumericTable::doAllocate);
+        (*(modelImplRef._probTbl))[resId].reset(probTbl);
+    }
+    return status;
 }
 
-services::Status ModelBuilder::addLeafNodeInternal(TreeId treeId, NodeId parentId, size_t position, size_t classLabel, NodeId& res)
+services::Status ModelBuilder::addLeafNodeInternal(const TreeId treeId, const NodeId parentId, const size_t position,
+    const size_t classLabel, NodeId & res)
 {
-    decision_forest::classification::internal::ModelImpl& modelImplRef = daal::algorithms::dtrees::internal::getModelRef<decision_forest::classification::internal::ModelImpl,ModelPtr>(_model);
-    return daal::algorithms::dtrees::internal::addLeafNodeInternal<size_t>(modelImplRef._serializationData, treeId, parentId, position, classLabel, res);
+    decision_forest::classification::internal::ModelImpl & modelImplRef =
+        daal::algorithms::dtrees::internal::getModelRef<decision_forest::classification::internal::ModelImpl, ModelPtr>(_model);
+    if (classLabel > _nClasses)
+    {
+        return services::Status(services::ErrorID::ErrorIncorrectParameter);
+    }
+    return daal::algorithms::dtrees::internal::addLeafNodeInternal<size_t>(modelImplRef._serializationData, treeId, parentId, position, classLabel,
+                                                                             res, modelImplRef._probTbl);
+}
+
+bool checkProba(const double * const proba, const size_t nClasses)
+{
+    double acc = 0.0;
+    const double epsilon = 10e-6;
+    for (size_t index = 0; index < nClasses; ++index)
+    {
+        if (proba[index] < 0.0 - epsilon || proba[index] > 1.0 + epsilon)
```

Comment:
Do we really wand to allow negative values for probabilities?
Looks strange: -0.000001 will pass this check for example.

Reply:
Accumulated double values of probability can be a little bit more than 1 or less than 0.